libgadget/fof.c

```diff
@@ -382,40 +373,96 @@ void fof_label_primary(void)
     myfree(FOF_PRIMARY_GET_PRIV(tw)->Head);
 }
 
+/* This function does the actual merging of the trees. It takes two head particles, h1 and h2,
+ * and merges 2 into h1 by setting Head[h2] = h1. It does this if and only if h1 < h2 (to avoid cycles)
+ * and it does it under a lock for h2, so that h2 cannot be merged elsewhere. h1 is not locked, to avoid deadlocking.
+ * It can thus cease to be the head of the tree, but this is fine, because it will never be merged into h2.
+ * If h1 > h2, the code swaps them and tries again.
+ * returns: 1 if a merge happened, 0 if it did not. No locks are held after return.*/
+static int
+try_merge_trees(int * h1, int * h2, MyIDType * subminID, MyIDType * subminIDTask, int * Head)
+{
+    /* We are about to change the HEAD of h2, so lock it.
+     * We need to make sure that h2 is really the head of the subtree,
+     * or not all particles will be linked.
+     * If this is not the case, unlock the particle, walk the tree again,
+     * and loop until we can get a lock.*/
+    while(1) {
+        /*Stop if we encountered h1 along the head of this tree: means this is already merged.*/
+        if(*h2 == -1)
+            return 0;
+        lock_particle(*h2);
+        if(Head[*h2] != *h2) {
+            unlock_particle(*h2);
+            *h2 = HEADa(*h1, *h2, Head);
+        }
+        else
+            break;
+    }
+    /*Get the current subtree properties: this is fine because we are under the lock.*/
```

Comment:
This amounts to finding a snapshot of a tree branch. After the lock is obtained no other thread will modify the branch. Therefore you can assert the MinID of the branch is accurate before you release the lock.

Reply:
I'm worried a bit about a thread modifying the branch after we release the lock and before we get the lock on h1 later. I *think* it is ok, because we only care about MinID for the new tree head, h1, and the other thread will modify that itself if needed. But I'm not sure.

Reply:
I think this is OK because the reduction of MinID is associative: as long as the MinID is reduced to the true head at some point then the MinID of the final head will be correct.

Reply:
Thanks, that sounds right.